libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -579,11 +579,14 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 					"(" <<
 					joinHumanReadable(args) <<
 					")\n";
+
+				m_context.functionGenerationQueue().push(functionDef);
```

Comment:
You only need to push the most derived one.

Maybe it would be cleaner to combine `virtualFunctionName` and pushing to the queue?

Reply:
If you always use `virtualFunctionName` is that also considering cases where you call directly a certain function like `BaseClass::myFunc()`?

Reply:
Good point. I'll change it to handle only the most derived one.

About `virtualFunctionName()`, yes, I think it would be cleaner since all the pushes would be in `IRGeneratorForStatements`. But I also think two separate calls are more readable. I would not expect something like `virtualFunctionName()` to have side-effects. So my suggestion would be to leave it as is. Or at least to change the name of the function if we combine it. What do you think?

Reply:
@Marenz I guess so, since `virtualFunctionName()` is called in `endVisit(FunctionCall)` for all internal calls, without discerning those two situations. But I'll test it with such a call just in case.

Reply:
@chriseth 
> You only need to push the most derived one.

Done. This is something that potentially breaks the generated code (the overriding function gets called but the code is only generated for the overridden one) so I added a new semantic test that would catch it.

> Maybe it would be cleaner to combine virtualFunctionName and pushing to the queue?

I moved it there but also changed the name to `enqueueVirtualFunctionForCodeGeneration()`. This actually fits quite well with the `enqueueFunctionForCodeGeneration()` helper I also added to the context - the helper pushes a function to the queue and returns its name so I can use it as a drop-in replacement for `generateFunction()`.